File: CarHomeWidget/src/main/java/com/anod/car/home/backup/ui/FragmentRestoreWidget.java
package com.anod.car.home.backup.ui;

import android.app.Activity;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.anod.car.home.R;
import com.anod.car.home.prefs.ConfigurationActivity;
import com.anod.car.home.backup.BackupCodeRender;
import com.anod.car.home.backup.BackupTask;
import com.anod.car.home.backup.GDriveBackup;
import com.anod.car.home.backup.PreferencesBackupManager;
import com.anod.car.home.backup.RestoreCodeRender;
import com.anod.car.home.backup.RestoreTask;
import info.anodsplace.android.log.AppLog;
import com.anod.car.home.utils.CheatSheet;
import com.anod.car.home.utils.DeleteFileTask;

import java.io.File;
import java.util.ArrayList;

import butterknife.Bind;
import butterknife.ButterKnife;

/**
 * @author algavris
 * @date 30/07/2016.
 */
public class FragmentRestoreWidget extends Fragment implements RestoreTask.RestoreTaskListener, DeleteFileTask.DeleteFileTaskListener,
        BackupTask.BackupTaskListener, GDriveBackup.Listener {

    public static FragmentRestoreWidget create(int appWidgetId) {
        FragmentRestoreWidget fragment = new FragmentRestoreWidget();
        Bundle args = new Bundle();
        args.putInt(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        fragment.setArguments(args);
        return fragment;
    }

    @Bind(android.R.id.list)
    ListView mListView;

    RestoreClickListener mRestoreListener;
    DeleteClickListener mDeleteListener;
    ExportClickListener mExportListener;

    int mAppWidgetId = AppWidgetManager.INVALID_APPWIDGET_ID;
    PreferencesBackupManager mBackupManager;
    RestoreAdapter mAdapter;
    GDriveBackup mGDriveBackup;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_restore_widget, container, false);
        ButterKnife.bind(this, view);

        mListView.setEmptyView(ButterKnife.findById(view, android.R.id.empty));
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mAppWidgetId = getArguments().getInt(AppWidgetManager.EXTRA_APPWIDGET_ID);
        mBackupManager = getRestoreFragment().getBackupManager();
        mGDriveBackup = new GDriveBackup(getActivity(), this);
        ((ConfigurationActivity) getActivity()).setActivityResultListener(mGDriveBackup);

        mRestoreListener = new RestoreClickListener();
        mDeleteListener = new DeleteClickListener();
        mExportListener = new ExportClickListener();

        mAdapter = new RestoreAdapter(getContext(), R.layout.fragment_restore_item, new ArrayList<File>());
        mListView.setAdapter(mAdapter);
        new FileListTask().execute(0);
        setHasOptionsMenu(true);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        if (!mGDriveBackup.isSupported())
        {
            menu.findItem(R.id.menu_download_from_cloud).setVisible(false);
        }
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_new_backup:
                createBackupNameDialog().show();
                return true;
            case R.id.menu_download_from_cloud:
                mGDriveBackup.download(mAppWidgetId);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent resultData) {

        if (requestCode == FragmentBackup.DOWNLOAD_MAIN_REQUEST_CODE
                && resultCode == Activity.RESULT_OK) {

            // The document selected by the user won't be returned in the intent.
            // Instead, a URI to that document will be contained in the return intent
            // provided to this method as a parameter.
            // Pull that URI using resultData.getData().
            if (resultData != null) {
                Uri uri = resultData.getData();
                AppLog.d("Uri: " + uri.toString());
                new RestoreTask(PreferencesBackupManager.TYPE_MAIN, mBackupManager, mAppWidgetId, FragmentRestoreWidget.this)
                        .execute(uri);

            }
        } else if (mGDriveBackup.checkRequestCode(requestCode)) {
            mGDriveBackup.onActivityResult(requestCode, resultCode, resultData);
        }
    }

    @Override
    public void onPause() {
        mGDriveBackup.disconnect();
        super.onPause();
    }

    private FragmentBackup getRestoreFragment() {
        return ((FragmentBackup) getParentFragment());
    }

    @Override
    public void onBackupPreExecute(int type) {
        getRestoreFragment().startRefreshAnim();
    }

    @Override
    public void onBackupFinish(int type, int code) {
        if (code == PreferencesBackupManager.RESULT_DONE) {
            new FileListTask().execute(0);
        }

        getRestoreFragment().stopRefreshAnim();
        int res = BackupCodeRender.render(code);
        Toast.makeText(getContext(), res, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onDeleteFileFinish(boolean success) {
        if (!success) {
            Toast.makeText(getContext(), R.string.unable_delete_file, Toast.LENGTH_SHORT).show();
        } else {
            new FileListTask().execute(0);
        }
    }

    static class ViewHolder {

        @Bind(android.R.id.title)
        TextView title;

        @Bind(android.R.id.text2)
        TextView text2;

        @Bind(R.id.apply_icon)
        ImageView apply;

        @Bind(R.id.delete_button)
        ImageView delete;

        @Bind(R.id.uploadMain)
        ImageView export;

        ViewHolder(View view) {
            ButterKnife.bind(this, view);
        }
    }

    private class RestoreAdapter extends ArrayAdapter<File> {

        private final int mResource;

        RestoreAdapter(Context context, int resource, ArrayList<File> items) {
            super(context, resource, items);
            mResource = resource;
        }

        @Override
        public View getView(int position, View view, ViewGroup parent) {
            ViewHolder holder;
            if (view != null) {
                holder = (ViewHolder) view.getTag();
            } else {
                LayoutInflater inflater = (LayoutInflater) getContext()
                        .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
                view = inflater.inflate(mResource, parent, false);
                holder = new ViewHolder(view);
                view.setTag(holder);
            }

            File entry = getItem(position);

            String name = entry.getName();
            String title = name.substring(0, name.lastIndexOf("."));
            holder.title.setTag(entry.getName());
            holder.title.setText(title);
            holder.title.setOnClickListener(mRestoreListener);

            String timestamp = DateUtils.formatDateTime(getContext(), entry.lastModified(), FragmentBackup.DATE_FORMAT);
            holder.text2.setText(timestamp);

            holder.apply.setTag(name);
            holder.apply.setOnClickListener(mRestoreListener);
            CheatSheet.setup(holder.apply);

            holder.delete.setTag(entry);
            holder.delete.setOnClickListener(mDeleteListener);
            CheatSheet.setup(holder.delete);

            if (mGDriveBackup.isSupported()) {
                holder.export.setTag(name);
                holder.export.setOnClickListener(mExportListener);
                CheatSheet.setup(holder.export);
            } else {
                holder.export.setVisibility(View.GONE);
            }

            view.setId(position);
            return view;
        }
    }

    class FileListTask extends AsyncTask<Integer, Void, File[]> {

        @Override
        protected void onPreExecute() {
            mAdapter.clear();
            mAdapter.notifyDataSetChanged();
        }

        protected File[] doInBackground(Integer... params) {
            return mBackupManager.getMainBackups();
        }

        protected void onPostExecute(File[] result) {
            if (result != null) {
                for (int i = 0; i < result.length; i++) {
                    mAdapter.add(result[i]);
                    mAdapter.notifyDataSetChanged();
                }
            }
        }
    }


    @Override
    public void onRestorePreExecute(int type) {
        getRestoreFragment().startRefreshAnim();
    }

    @Override
    public void onRestoreFinish(int type, int code) {
        getRestoreFragment().stopRefreshAnim();
        int res = RestoreCodeRender.render(code);
        Toast.makeText(getContext(), res, Toast.LENGTH_SHORT).show();
    }

    class RestoreClickListener implements View.OnClickListener {

        @Override
        public void onClick(View v) {

            Uri uri = Uri.fromFile(mBackupManager.getBackupWidgetFile((String) v.getTag()));
            new RestoreTask(PreferencesBackupManager.TYPE_MAIN, mBackupManager, mAppWidgetId,
                    FragmentRestoreWidget.this)
                    .execute(uri)
            ;
        }
    }

    class DeleteClickListener implements View.OnClickListener {

        @Override
        public void onClick(View v) {
            File file = (File) v.getTag();
            new DeleteFileTask(FragmentRestoreWidget.this).execute(file);
        }
    }

    AlertDialog createBackupNameDialog() {
        String defaultFilename = "widget-" + mAppWidgetId;
        // This example shows how to add a custom layout to an AlertDialog
        LayoutInflater factory = LayoutInflater.from(getContext());
        final View textEntryView = factory.inflate(R.layout.backup_dialog_enter_name, null);
        final EditText backupName = (EditText) textEntryView.findViewById(R.id.backup_name);
        backupName.setText(defaultFilename);

        return new AlertDialog.Builder(getContext())
                .setTitle(R.string.backup_current_widget)
                .setView(textEntryView)
                .setPositiveButton(R.string.backup_save, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int whichButton) {
                        String filename = backupName.getText().toString();
                        if (!filename.equals("")) {
                            new BackupTask(PreferencesBackupManager.TYPE_MAIN, mBackupManager,
                                    mAppWidgetId, FragmentRestoreWidget.this).execute(filename);
                        }
                    }
                })
                .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int whichButton) {
                        //Nothing
                    }
                }).create()
                ;
    }

    class ExportClickListener implements View.OnClickListener {

        @Override
        public void onClick(View v) {
            String name = (String) v.getTag();
            File file = mBackupManager.getBackupWidgetFile(name);
            mGDriveBackup
                    .upload("car-" + name, file);
        }
    }

    @Override
    public void onGDriveActionStart() {
        getRestoreFragment().startRefreshAnim();
    }

    @Override
    public void onGDriveDownloadFinish() {
        onRestoreFinish(PreferencesBackupManager.TYPE_MAIN, PreferencesBackupManager.RESULT_DONE);
    }

    @Override
    public void onGDriveUploadFinish() {
        onBackupFinish(PreferencesBackupManager.TYPE_MAIN, PreferencesBackupManager.RESULT_DONE);
    }

    @Override
    public void onGDriveError() {
        onRestoreFinish(PreferencesBackupManager.TYPE_MAIN, PreferencesBackupManager.ERROR_UNEXPECTED);
    }
}

File: CarHomeWidget/src/main/java/com/anod/car/home/backup/ui/FragmentRestoreWidget.kt
package com.anod.car.home.backup.ui

import android.app.Activity
import android.app.Application
import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.core.content.FileProvider
import androidx.core.net.toUri
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.observe
import com.anod.car.home.BuildConfig
import com.anod.car.home.R
import com.anod.car.home.backup.BackupCodeRender
import com.anod.car.home.backup.PreferencesBackupManager
import com.anod.car.home.backup.RestoreCodeRender
import com.anod.car.home.utils.*
import info.anodsplace.framework.AppLog
import info.anodsplace.framework.app.DialogCustom
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

private class RestoreWidgetViewModel(application: Application) : RestoreViewModel(application) {
    init {
        type = PreferencesBackupManager.TYPE_MAIN
    }

    fun backup(filename: String) {
        backup(backupManager.getBackupWidgetFile(filename).toUri())
    }

    fun upload(name: String) {
        val file = backupManager.getBackupWidgetFile(name)
        upload(name, file)
    }

    override suspend fun refreshFiles(): List<File> = withContext(Dispatchers.IO) {
        backupManager.mainBackups.toList()
    }
}

/**
 * @author algavris
 * @date 30/07/2016.
 */
class FragmentRestoreWidget : Fragment() {

    private var listView: ListView? = null

    private var appWidgetId = AppWidgetManager.INVALID_APPWIDGET_ID
    private val viewModel: RestoreWidgetViewModel by viewModels()
    private val adapter: RestoreAdapter by lazy {
        RestoreAdapter(context!!,
                View.OnClickListener { viewModel.restore(it.tag as Uri) },
                View.OnClickListener { viewModel.delete(it.tag as File) },
                View.OnClickListener { viewModel.upload(it.tag as String) }
        )
    }

    private val restoreFragment: FragmentBackup
        get() = parentFragment as FragmentBackup

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_restore_widget, container, false)
        listView = view.findViewById(android.R.id.list)
        listView!!.emptyView = view.findViewById(android.R.id.empty)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel.appWidgetId = arguments!!.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID)
        viewModel.backupManager = restoreFragment.backupManager

        viewModel.uploadEvent.observe(this) {
            upload(it.first, it.second)
        }

        viewModel.restoreEvent.observe(this) { code ->
            when (code) {
                PreferencesBackupManager.NO_RESULT -> restoreFragment.startRefreshAnim()
                else -> {
                    restoreFragment.stopRefreshAnim()
                    Toast.makeText(context, RestoreCodeRender.render(code), Toast.LENGTH_SHORT).show()
                }
            }
        }

        viewModel.backupEvent.observe(this) { code ->
            when (code) {
                PreferencesBackupManager.NO_RESULT -> restoreFragment.startRefreshAnim()
                else -> {
                    restoreFragment.stopRefreshAnim()
                    Toast.makeText(context, BackupCodeRender.render(code), Toast.LENGTH_SHORT).show()
                }
            }
        }

        viewModel.deleteEvent.observe(this) { success ->
            if (!success) {
                Toast.makeText(context, R.string.unable_delete_file, Toast.LENGTH_SHORT).show()
            }
        }

        viewModel.files.observe(this) {
            adapter.clear()
            adapter.addAll(it)
            adapter.notifyDataSetChanged()
        }

        listView!!.adapter = adapter
        if (AppPermissions.isGranted(context!!, ReadExternalStorage)) {
            viewModel.loadFiles()
        } else {
            AppPermissions.request(this, arrayOf(ReadExternalStorage, WriteExternalStorage), requestList)
        }
        setHasOptionsMenu(true)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_new_backup -> {
                backup()
                return true
            }
            R.id.menu_download_from_cloud -> {
                download()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    fun backup() {
        if (AppPermissions.isGranted(context!!, WriteExternalStorage)) {
            createBackupNameDialog("widget-$appWidgetId").show()
        } else {
            AppPermissions.request(this, arrayOf(ReadExternalStorage, WriteExternalStorage), requestBackup)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        AppPermissions.checkResult(requestCode, grantResults, requestBackup) {
            when (it) {
                is Granted -> backup()
                is Denied -> Toast.makeText(context, "Permissions are required", Toast.LENGTH_SHORT).show()
            }
        }

        AppPermissions.checkResult(requestCode, grantResults, requestList) {
            when (it) {
                is Granted -> viewModel.loadFiles()
                is Denied -> Toast.makeText(context, "Permissions are required", Toast.LENGTH_SHORT).show()
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, result: Intent?) {
        if (requestCode == requestUpload) {
            if (resultCode == Activity.RESULT_OK) {
                val uri = result!!.data!!
                AppLog.d("Uri: $uri")
                viewModel.backup(uri)
            }
        } else if (requestCode == requestDownload) {
            if (resultCode == Activity.RESULT_OK) {
                val uri = result!!.data
                AppLog.d("Uri: " + uri!!.toString())
                viewModel.backup(uri)
            }
        }
    }

    class ViewHolder(view: View) {
        var title: TextView = view.findViewById(android.R.id.title)
        var text2: TextView = view.findViewById(android.R.id.text2)
        var apply: ImageView = view.findViewById(R.id.apply_icon)
        var delete: ImageView = view.findViewById(R.id.delete_button)
        var export: ImageView = view.findViewById(R.id.uploadMain)
    }

    private class RestoreAdapter constructor(
            context: Context,
            private val restoreListener: View.OnClickListener,
            private val deleteListener: View.OnClickListener,
            private val exportListener: View.OnClickListener)
        : ArrayAdapter<File>(context, R.layout.fragment_restore_item) {

        override fun getView(position: Int, view: View?, parent: ViewGroup): View {
            var itemView = view
            val holder: ViewHolder
            if (itemView != null) {
                holder = itemView.tag as ViewHolder
            } else {
                val inflater = context
                        .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
                itemView = inflater.inflate(R.layout.fragment_restore_item, parent, false)
                holder = ViewHolder(itemView)
                itemView!!.tag = holder
            }

            val entry = getItem(position)

            val name = entry!!.name
            val title = name.substring(0, name.lastIndexOf("."))
            holder.title.tag = entry.toUri()
            holder.title.text = title
            holder.title.setOnClickListener(restoreListener)

            val timestamp = DateUtils.formatDateTime(context, entry.lastModified(), FragmentBackup.DATE_FORMAT)
            holder.text2.text = timestamp

            holder.apply.tag = entry.toUri()
            holder.apply.setOnClickListener(restoreListener)
            CheatSheet.setup(holder.apply)

            holder.delete.tag = entry
            holder.delete.setOnClickListener(deleteListener)
            CheatSheet.setup(holder.delete)

            holder.export.tag = name
            holder.export.setOnClickListener(exportListener)
            CheatSheet.setup(holder.export)

            itemView.id = position
            return itemView
        }
    }

    private fun createBackupNameDialog(currentName: String): AlertDialog {
        // This example shows how to add a custom layout to an AlertDialog
        return DialogCustom(context!!, 0, R.string.save_widget, R.layout.backup_dialog_enter_name) {
            view, builder ->

            val backupName = view.findViewById<EditText>(R.id.backup_name)
            backupName.setText(currentName)

            builder.setPositiveButton(R.string.backup_save) { _, _ ->
                val filename = backupName.text.toString()
                if (filename.isNotBlank()) {
                    viewModel.backup(filename)
                }
            }

            builder.setNegativeButton(android.R.string.cancel) { _, _ -> }

        }.create()
    }

    private fun download() {
        startActivityForResultSafetly(Intent().forOpenBackup(), requestDownload)
    }

    private fun upload(filename: String, source: File) {
        val uri = FileProvider.getUriForFile(context!!.applicationContext, AUTHORITY, source)
        startActivityForResultSafetly(Intent().forCreateBackup(filename, uri), requestUpload)
    }

    companion object {
        const val requestUpload = 123
        const val requestDownload = 124
        const val requestBackup = 125
        const val requestList = 126

        const val AUTHORITY = BuildConfig.APPLICATION_ID + ".fileprovider"

        fun create(appWidgetId: Int): FragmentRestoreWidget {
            val fragment = FragmentRestoreWidget()
            val args = Bundle()
            args.putInt(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
            fragment.arguments = args
            return fragment
        }
    }
}